### PR TITLE
SAR - Resubmission Question

### PR DESCRIPTION
**Pull request by keeysnc**
### Description
Added resubmission field in SAR. The question field is visible:
1. If report is SAR
2. If a report is in  'In Revision'
3. If a report has been submitted more than 1 time and is being re-revised

### Related ticket(s)
https://jiraent.cms.gov/browse/CMDCT-3015

https://github.com/Enterprise-CMCS/macpro-mdct-mfp/assets/14514294/e9b898ae-80b1-4ea0-a5a8-13b338f1d926

---
### How to test
1. Create a WP as a state user and submit it

2. Approve WP report as an admin

3. Create a Sar with state user ( field should be hidden in general information)

4. Submit SAR

5. Log in as admin and unlock the report ( field should be visible after unlocking)

6. Log back in as state user and you ( field should be visible and cleared of any value)

7. Fill out the field and resubmit the SAR 

8. Log back in as an admin and the field should display new revision text in field

9. If you unlock the SAR again as an admin and return to the SAR as a state user the 'resubmission' field should reset to an empty field

---
### Author checklist

- [ ] I have performed a self-review of my code
- [ ] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

**Review by karla-vm — APPROVED** [review]
LGTM, awesome work on this one, it's a tricky use-case! 🥇

**Review by gmrabian — APPROVED** [review]

**Review by gmrabian — COMMENTED** [review]
If you submit as a state user, before unlocking as an admin, the revision question shows up with "N/A" if you go back in to view the report. Is this ok? Seems like it should only show up after it's been unlocked the first time, though maybe not a big deal. I'm just imagining a state user might submit and then go to print the report and a question about resubmission shows up at the top that they aren't expecting

**Comment by keeysnc** [comment]
> If you submit as a state user, before unlocking as an admin, the revision question shows up with "N/A" if you go back in to view the report. Is this ok? Seems like it should only show up after it's been unlocked the first time, though maybe not a big deal. I'm just imagining a state user might submit and then go to print the report and a question about resubmission shows up at the top that they aren't expecting

Hm, I can take a quick look at this.

**Comment by keeysnc** [comment]
> If you submit as a state user, before unlocking as an admin, the revision question shows up with "N/A" if you go back in to view the report. Is this ok? Seems like it should only show up after it's been unlocked the first time, though maybe not a big deal. I'm just imagining a state user might submit and then go to print the report and a question about resubmission shows up at the top that they aren't expecting

Updated.

**Review by karla-vm — APPROVED** [review]

File: services/ui-src/src/components/reports/StandardReportPage.tsx [hunk 1]
```diff
@@ -58,6 +59,42 @@ export const StandardReportPage = ({ route, validateOnRender }: Props) => {
...
+    ) {
+      return formDataCopy;
+    } else {
+      const removeResubmissionDataFields: any[] = formDataFields.filter(
+        (data: AnyObject) =>
+          data.id !== "generalInformation_resubmissionInformation"
+      );
+      formDataCopy.fields = [...removeResubmissionDataFields];
+
+      return formDataCopy;
```
Comment (gmrabian) [inline services/ui-src/src/components/reports/StandardReportPage.tsx:94]: the `else` and the first `if` now have the same logic, right?

Comment (gmrabian) [inline services/ui-src/src/components/reports/StandardReportPage.tsx:94]: the else and the first if now have the same logic, right?

File: services/ui-src/src/components/reports/StandardReportPage.tsx [hunk 2]
```diff
@@ -58,6 +59,42 @@ export const StandardReportPage = ({ route, validateOnRender }: Props) => {
     navigate(nextRoute);
   };
 
+  const filterResubmissionData = (formData: FormJson) => {
+    //make deep copies of formData
+    const formDataCopy = JSON.parse(JSON.stringify(formData));
+    const formDataFields = JSON.parse(JSON.stringify(formData?.fields));
```
Comment (gmrabian) [inline services/ui-src/src/components/reports/StandardReportPage.tsx:65]: suggestion: there's a native js function for this now called `structuredClone()`! does basically the same thing as json.parse(json.stringify)

**Review by gmrabian — COMMENTED** [review]
Form logic works as expected! Can simplify the code a little I think

File: services/ui-src/src/components/reports/StandardReportPage.tsx (hunk 1 above)
Comment (keeysnc) [reply services/ui-src/src/components/reports/StandardReportPage.tsx:94]: Same logic but different condition. I could put it in a function..

File: services/ui-src/src/components/reports/StandardReportPage.tsx (hunk 2 above)
Comment (keeysnc) [reply services/ui-src/src/components/reports/StandardReportPage.tsx:65]: oh nice! Thanks

**Review by gmrabian — APPROVED** [review]

**Review by gmrabian — CHANGES_REQUESTED** [review]
re: tests failing on structuredClone - oh yeah older versions of jest don't have this yet :/ we encountered this elsewhere in MFP.

If you add this to the test file it will fix it. We do that elsewhere in MFP
```
global.structuredClone = jest.fn((val) => {
  return JSON.parse(JSON.stringify(val));
});
```

**Review by gmrabian — APPROVED** [review]
I think we got there 💪

**Review by karla-vm — APPROVED** [review]

**Pull request merged**